raiden/network/transport/matrix.py

```diff
@@ -573,28 +644,65 @@ def _send_queued_messages(
 
     def _send_with_retry(
         self,
-        receiver_address: Address,
-        async_result: AsyncResult,
-        data: str,
+        queue_identifier: QueueIdentifier,
+        message: Message,
     ):
+        data = json.dumps(message.to_dict())
+        message_id = message.message_identifier
+        receiver_address = queue_identifier.recipient
+        reachable = {UserPresence.ONLINE, UserPresence.UNAVAILABLE}
+
         def retry():
-            if not self._running:
-                return
             timeout_generator = udp_utils.timeout_exponential_backoff(
                 self._config['retries_before_backoff'],
                 self._config['retry_interval'],
                 self._config['retry_interval'] * 10,
             )
-            while async_result.value is None:
-                self._send_immediate(receiver_address, data)
-                gevent.sleep(next(timeout_generator))
+            for delay in timeout_generator:
+                state = self._address_to_presence.get(receiver_address)
+                if state in reachable:
+                    self._send_raw(receiver_address, data)
```

Comment:
This is a blocking call, correct?

The `if self._stop_event.wait(delay):` bellow will not be sufficient to exit the task on a timely manner. Is the kill introduced in raiden-libs used to exit this?

Reply:
Yes, `_send_raw` blocks, but should be very quick (only request send time), so why `_stop_event.wait(delay)` isn't enough? The `Timeout` on `stop` waiting for these greenlets should ensure it won't block forever